### ISSUE TITLE
Train a transformer world model that can explicitly retrieve past experiences

### ISSUE THREAD

**Issue by mmatena** [mmatena/comp_755_project#6]
- The basic idea is to make a model that predicts the next state and the reward from the transition given a history of previous states, actions,and rewards and also given the action taken from that state.
- This is very similar to a LM.
- The goal is to see whether adding some sort of explicit retrieval mechanism over past experiences can improve perplexity on this task.
    - First is to try something like the nearest neighbor look up as it is easy and doesn’t involve additional training.
    - Then I can look into stuff like REALM.
- Architecture is a transformer with a relatively short sequence length. I should experiment with that from 16 to 128.
- Don’t count loss on the first few (maybe 4?) steps as the model might not be able to compute stuff like velocity given only a few frames.
- Also get a validation set, especially when evaluating using an explicit lookup over the train data.

**Comment by mmatena** [comment]
It looks like the transformer I trained outputs a constant value regardless of the input. It's possible this happened due to the MSE loss. I'll try training a transformer that outputs a mixture of Gaussians with a temperature like the World Models paper.

**Comment by mmatena** [comment]
I'll also try having it predict the difference between the current state and the next state. Or use the difference as the returned values from the knn lookup.

**Comment by mmatena** [comment]
It looks like the issue was having the transformer too large. I was using the BERT base config. Making the model shallower and decreasing the size helped. I'm guessing the complexity of the world model, at least for car racing, is less than for natural text, so smaller models shouldn't be much of a problem.

As a side note, I'm pretty sure there's a big difference in terms whether the layer norms go before/after something (I forgot the specifics). One version makes it hard to train deep models, necessitating stuff like [transparent attention](https://arxiv.org/pdf/1808.07561.pdf). With the other version, we can train deeper model without issue. I know t5 used the version that works better.

If we find benefit from training deeper models, then we should probably make sure we are using the correct version.

**Comment by mmatena** [comment]
I did a bit more experimenting. It seems to have more to do with the width of the model rather than the depth. It could just be an optimization issue or something more fundamental. I'll be going with `d_model = 256` for now.

**Comment by mmatena** [comment]
It looks like another issue was that I was using `tf.distribute.MirroredStrategy()` for training. I did some experiments where I trained with the same setup but either did or not use the MirroredStrategy. The ones with the MirroredStrategy led to constant values while the ones without were OK.

The point of the MirroredStrategy was to distribute training across GPUs. I tried without it and, at least in this case, it was just as fast. I haven't look into the solution for when we need a MirroredStrategy, but one thing to try is to try to create the model and load the weights within the `with mirrored_strategy.scope():` block.

I also think having the model too wide was still an issue in my intial attempts.

**Comment by mmatena** [comment]
I tried training such a model for 100k steps with a batch size of 128. It still returned roughly a constant output. I'm thinking of adding a loss term that ensures the variance of the output is equal to the variance of the targets in each sequence.
